### PR TITLE
Termination Notification Configurable Action

### PR DESCRIPTION
**Pull request by mello7tre**
# Issue Type

- Feature Pull Request

## Summary

Idea:
At the moment, when a termination notification message is received,
we are only supporting detach as action on the terminating spot instance.

Detaching an instance do not trigger Lifecycle Hook (LH).

This patch add a new option "termination_notification_action" with
relative CloudFormation parameter "TerminationNotificationAction" that
support the options:
```
detach
	current behaviour - no LH Triggered
terminate
	new behaviour - spot instance is terminated without decrementing
	desired capacity, so that ASG spawn a new instance (same as detach) - LH
	Triggered.
auto
	new behaviour - if ASG has a LH for Terminate Lifecycle
	Transition, action is "terminate", if not action is detach.
```
The option is global for all ASG.
But using "auto", behaviour is based on the presence of LH, this way it
can act in different ways.

## Code contribution checklist

1. [ ] The contribution fixes a single existing github issue, and it is linked
   to it.
1. [x] The code is as simple as possible, readable and follows the idiomatic Go
  [guidelines](https://golang.org/doc/effective_go.html).
1. [x] All new functionality is covered by automated test cases so the overall
  test coverage doesn't decrease.
1. [x] No issues are reported when running `make full-test`.
1. [x] Functionality not applicable to all users should be configurable.
1. [x] Configurations should be exposed through Lambda function environment
   variables which are also passed as parameters to the
   [CloudFormation](https://github.com/cristim/autospotting/blob/master/cloudformation/stacks/AutoSpotting/template.yaml)
   and
   [Terraform](https://github.com/autospotting/terraform-aws-autospotting/main.tf)
   stacks defined as infrastructure code.
1. [ ] Global configurations set from the infrastructure stack level should also
   support per-group overrides using tags.
1. [x] Tags names and expected values should be similar to the other existing
   configurations.
1. [x] Both global and tag-based configuration mechanisms should be tested and
  proven to work using log output from various test runs.
1. [x] The logs should be kept as clean as possible (use log levels as
  appropriate) and formatted consistently to the existing log output.
1. [ ] The documentation is updated to cover the new behavior, as well as the
   new configuration options for both stack parameters and tag overrides.
1. [ ] A code reviewer reproduced the problem and can confirm the code
  contribution actually resolves it.

### CODE REVIEW

File: cloudformation/stacks/AutoSpotting/template.yaml [hunk 1]
```diff
@@ -40,6 +40,17 @@
         "Instance termination method.  Must be one of 'autoscaling' (default) or
         'detach' (compatibility mode, not recommended)"
       Type: "String"
+    TerminationNotificationAction:
+      AllowedValues:
+        - "auto"
+        - "detach"
+        - "terminate"
+      Default: "terminate"
```
Comment (cristim) [inline cloudformation/stacks/AutoSpotting/template.yaml]: I would set the default to "auto"

File: core/spot_termination.go [hunk 2]
```diff
@@ -101,3 +116,57 @@ func (s *SpotTermination) getAsgName(instanceID *string) (string, error) {
...
+			s.terminateInstance(instanceID, asgName)
+		} else {
+			s.detachInstance(instanceID, asgName)
+		}
+	}
+
+	return nil
+}
+
+func (s *SpotTermination) asgHasHook(autoScalingGroupName *string) bool {
```
Comment (cristim) [inline core/spot_termination.go]: I would rename this to `asgHasTerminationLifecycleHook` or if we keep the name as it is, we can make it take the hook type as an additional parameter.

File: cloudformation/stacks/AutoSpotting/template.yaml (hunk 1 above)
Comment (mello7tre) [reply cloudformation/stacks/AutoSpotting/template.yaml]: Sure, that is the right one!
Just forgot to make the change.

File: core/spot_termination.go (hunk 2 above)
Comment (mello7tre) [reply core/spot_termination.go]: Ok. As the only LF that make sense, in that contest, is the Termination one i will rename it as your suggestion.

**Pull request merged**